### ISSUE TITLE
Error when creating new product

### ISSUE THREAD

**Issue by flgatormike** [nozzlegear/ShopifySharp#723]
This library is great. I can get products, orders, customers, and even update a product, but when I try to create a product using CreateAsync, I get:
"Object reference not set to an instance of an object"
Any idea what the problem might be?
Thanks!

**Comment by nozzlegear** [comment]
Thanks for the bug report @flgatormike! This package has unit tests that run with every new release, including ones for creating a new product, and they don't seem to have this issue. Could you possibly post the full exception with stack trace so I can take a look? It might also help if you can post the code you're using to create a product.

**Comment by flgatormike** [comment]
Thanks for the reply! I am trying to create a product by getting a product and just changing the id and title and then creating the product. I have also tried manually building the product. The stack trace is below but it looks like it doesn't step into the relevant code. Any idea what I might be doing wrong?

                var product = await GetProduct(7550715560151);
                product.Id = 7550715560133;
                product.Title = "TEST";
                product = await service.CreateAsync(product);


   at ShopifySharp.ShopifyService.<>c__DisplayClass27_0`1.<<ExecuteRequestAsync>b__0>d.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at ShopifySharp.DefaultRequestExecutionPolicy.<Run>d__0`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at ShopifySharp.ShopifyService.<ExecuteRequestAsync>d__27`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ShopifySharp.ProductService.<CreateAsync>d__5.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at qb_shopify.Main.<CreateProducts>d__22.MoveNext() in C:\_projects\emustiga\qb-shopify\Main.cs:line 368

**Comment by flgatormike** [comment]
I also just tried copying the code in the unit test and get the same exception:

        public async Task<Product> CreateTestProduct(ProductCreateOptions options = null)
        {
            var service = new ProductService(domain, accessToken);
            var obj = await service.CreateAsync(new Product()
            {
                Title = "Test Product",
                Vendor = "Test Vendor",
                BodyHtml = "<p>Test Product</p>",
                ProductType = "Product",
                Handle = Guid.NewGuid().ToString(),
                Images = new List<ProductImage>
                {
                    new ProductImage
                    {
                        Attachment = "R0lGODlhAQABAIAAAAAAAAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
                    }
                },
            }, options);

            return obj;
        }

**Comment by flgatormike** [comment]
Hmm, just tested and the ShopifyException isn't thrown, just the standard Exception

**Comment by nozzlegear** [comment]
Something else to try: I know you've copied the code from the test project and had it break there too, but in the other sample code you provided above it looks like you've got a method named `GetProduct` and you're using that base product data to create a new one. What happens if you use the `productService.GetAsync` instead of your custom method? Does that throw the same error when you try to create with it?

```cs
var product = service.GetAsync(7550715560151);
product.Id = 7550715560133;
product.Title = "TEST";
product = await service.CreateAsync(product);
```

**Comment by flgatormike** [comment]
so here is where I am getting the error. rawResult looks like it is valid json, but I am getting an error on deserialization: 
![image](https://user-images.githubusercontent.com/47644612/158269343-df36b5d2-edb5-46a5-852d-772b96d8038c.png)

**Comment by nozzlegear** [comment]
Thanks! So it is the response json causing the problem. Could you post that raw result value here? I’ll run it in a unit test and see what happens.

--
Josh Harms

**Comment by flgatormike** [comment]
Yes, the problem in the json is the root element. The root of the json "product" but the code is looking for "products" with an s.. not sure if this mismatch is a problem with the shopify store or my code?

**Comment by nozzlegear** [comment]
That would definitely be the source of the problem. The `ProductService.CreateAsync` method _should_ be using the string `"product"` as the root. I'm unable to locate any point where it was using `"products"` in the git history, but I might just be overlooking it. Are you overriding the method at all with your own implementation? If not, have you tried updating to the latest version of ShopifySharp to see if that fixes the problem?

**Comment by flgatormike** [comment]
"products" (with the s) is coming back from shopify, so maybe this is a problem with the store configuration?

**Comment by flgatormike** [comment]
Below is the code I am using to create a test product, which makes a call to executerequestasync which returns the json below which is a list of products and that is where I am getting the error. Not sure why I am getting a list of products here, and that is where I am getting the error...


                var product = await service.CreateAsync(new Product()
                {
                    Title = "Test Product",
                    Vendor = "Test Vendor",
                    BodyHtml = "<p>Test Product</p>",
                    ProductType = "Product",
                    Handle = Guid.NewGuid().ToString(),
                    Images = new List<ProductImage>
                    {
                        new ProductImage
                        {
                            Attachment = "R0lGODlhAQABAIAAAAAAAAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
                        }
                    },
                }, options);






{"products":[{"id":7553612677335,"title":"Case OPC-8301-3","body_html":"\u003cspan id=\"ctl00_ContentPlaceHolder1_lblDescription\" data-mce-fragment=\"1\"\u003eSemi-rigid sun case w\/zipper. Red and black. Oval shape. \u003cbr data-mce-fragment=\"1\"\u003eDimensions: 17 x 3in \/ 18 x 18 cm \u003cbr data-mce-fragment=\"1\"\u003eWeight: 2.00 oz \/ 56.7 g\u003c\/span\u003e","vendor":"optiusa","product_type":"","created_at":"2022-02-21T15:38:24-05:00","handle":"case-opc-8301-3","updated_at":"2022-03-03T15:57:59-05:00","published_at":"2022-02-22T10:48:24-05:00","template_suffix":"opti-cases","status":"active","published_scope":"web","tags":"","admin_graphql_api_id":"gid:\/\/shopify\/Product\/7553612677335","variants":[{"id":42431478235351,"product_id":7553612677335,"title":"Blue","price":"1.95","sku":"1","position":1,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Blue","option2":null,"option3":null,"created_at":"2022-02-21T15:38:24-05:00","updated_at":"2022-02-21T15:50:45-05:00","taxable":true,"barcode":"","grams":57,"image_id":36673401192663,"weight":2.0,"weight_unit":"oz","inventory_item_id":44525932183767,"inventory_quantity":1,"old_inventory_quantity":1,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42431478235351"},{"id":42431478268119,"product_id":7553612677335,"title":"Gray","price":"1.95","sku":"1","position":2,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Gray","option2":null,"option3":null,"created_at":"2022-02-21T15:38:24-05:00","updated_at":"2022-02-21T15:50:56-05:00","taxable":true,"barcode":"","grams":57,"image_id":36673401454807,"weight":2.0,"weight_unit":"oz","inventory_item_id":44525932216535,"inventory_quantity":1,"old_inventory_quantity":1,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42431478268119"},{"id":42431478300887,"product_id":7553612677335,"title":"Red","price":"1.95","sku":"1","position":3,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Red","option2":null,"option3":null,"created_at":"2022-02-21T15:38:24-05:00","updated_at":"2022-02-21T15:50:51-05:00","taxable":true,"barcode":"","grams":57,"image_id":36673401520343,"weight":2.0,"weight_unit":"oz","inventory_item_id":44525932249303,"inventory_quantity":1,"old_inventory_quantity":1,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42431478300887"}],"options":[{"id":9613111230679,"product_id":7553612677335,"name":"Color","position":1,"values":["Blue","Gray","Red"]}],"images":[{"id":36673401192663,"product_id":7553612677335,"position":1,"created_at":"2022-02-21T15:38:38-05:00","updated_at":"2022-02-21T15:38:38-05:00","alt":null,"width":1200,"height":600,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/product_OPC-8301-1-1.jpg?v=1645475918","variant_ids":[42431478235351],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36673401192663"},{"id":36673401454807,"product_id":7553612677335,"position":2,"created_at":"2022-02-21T15:38:48-05:00","updated_at":"2022-02-21T15:38:48-05:00","alt":null,"width":1200,"height":600,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/product_OPC-8301-2-1.jpg?v=1645475928","variant_ids":[42431478268119],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36673401454807"},{"id":36673401520343,"product_id":7553612677335,"position":3,"created_at":"2022-02-21T15:38:56-05:00","updated_at":"2022-02-21T15:38:56-05:00","alt":null,"width":1200,"height":600,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/product_OPC-8301-3-1.jpg?v=1645475936","variant_ids":[42431478300887],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36673401520343"}],"image":{"id":36673401192663,"product_id":7553612677335,"position":1,"created_at":"2022-02-21T15:38:38-05:00","updated_at":"2022-02-21T15:38:38-05:00","alt":null,"width":1200,"height":600,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/product_OPC-8301-1-1.jpg?v=1645475918","variant_ids":[42431478235351],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36673401192663"}},{"id":7553620082903,"title":"CX 800 Auto-Refractor-Keratometer","body_html":"\u003cp\u003e\u003cspan style=\"text-decoration: underline;\"\u003e\u003cstrong\u003eRODENSTOCK INSTRUMENTS\u003c\/strong\u003e\u003c\/span\u003e\u003c\/p\u003e\n\u003cp\u003e\u003cem\u003e- German manufacture -\u003c\/em\u003e\u003c\/p\u003e\n\u003cp\u003e \u003c\/p\u003e\n\u003cp\u003e\u003cspan data-mce-fragment=\"1\"\u003eTechnical specifications:\u003c\/span\u003e\u003c\/p\u003e\n\u003cp\u003e \u003cspan id=\"ctl00_ContentPlaceHolder1_lblDescription\" data-mce-fragment=\"1\"\u003eExtremely fast operation\u003cbr data-mce-fragment=\"1\"\u003ePupil + cornea ø measurement\u003cbr data-mce-fragment=\"1\"\u003eColour touch screen\u003cbr data-mce-fragment=\"1\"\u003eAuto measurement\u003cbr data-mce-fragment=\"1\"\u003eCentral K-values\u003c\/span\u003e\u003c\/p\u003e\n\u003cp\u003e\u003cspan id=\"ctl00_ContentPlaceHolder1_lblDescription\" data-mce-fragment=\"1\"\u003e\u003ca href=\"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/files\/Manual_570-914.pdf?v=1645555139\" title=\"Download Brochure\"\u003eDownload Brochure\u003c\/a\u003e\u003cbr data-mce-fragment=\"1\"\u003e\u003cbr data-mce-fragment=\"1\"\u003eThe Rodenstock Refractor-Keratometers are an integral part of modern eye diagnostic. Highly accurate measurements combined with the very short examination time and easy handling makes working with the Rodenstock CX 800 professional and quick. You can operate and align the CX 800 through the combination of joystick and touch screen – all this in a very short time.\u003cbr data-mce-fragment=\"1\"\u003e\u003cbr data-mce-fragment=\"1\"\u003eColour touch screen\u003cbr data-mce-fragment=\"1\"\u003eThe 5.7“ colour touch screen is used as operating monitor as\u003cbr data-mce-fragment=\"1\"\u003ewell as for displaying all measured values. All commands can be\u003cbr data-mce-fragment=\"1\"\u003edone via touch screen.\u003cbr data-mce-fragment=\"1\"\u003e\u003cbr data-mce-fragment=\"1\"\u003ePupil + cornea ø measurement\u003cbr data-mce-fragment=\"1\"\u003eOnce you have captured your patients’ eye, you can set the pupil and cornea measurement bars to individual diameters. These\u003cbr data-mce-fragment=\"1\"\u003evalues will also be stored, displayed and printed.\u003cbr data-mce-fragment=\"1\"\u003e\u003cbr data-mce-fragment=\"1\"\u003eAuto measurement\u003cbr data-mce-fragment=\"1\"\u003eThe handling of the CX 800 is very easy. By aligning the optical\u003cbr data-mce-fragment=\"1\"\u003ehead towards the patients’ eye, the measurement is automatically taken over by the instrument itself. \u003cbr data-mce-fragment=\"1\"\u003e\u003cbr data-mce-fragment=\"1\"\u003eCentral K-values\u003cbr data-mce-fragment=\"1\"\u003eThe CX 800 provides you with the measured central (ø 3 mm)\u003cbr data-mce-fragment=\"1\"\u003eKeratometer readings within a second. You can take these readings from a cornea or the back surfaces of a RGP contact lens.\u003cbr data-mce-fragment=\"1\"\u003e\u003cbr data-mce-fragment=\"1\"\u003eSPECIFICATIONS\u003cbr data-mce-fragment=\"1\"\u003eRrefractive power measurement\u003cbr data-mce-fragment=\"1\"\u003eSpherical refractive power (S)\u003cbr data-mce-fragment=\"1\"\u003eMeasurement range .............. -25.00 D to +22.00 D\u003cbr data-mce-fragment=\"1\"\u003e (at VD = 12.0 mm)\u003cbr data-mce-fragment=\"1\"\u003eDisplay unit ........................... 0.01 D \/ 0.12 D \/ 0.25 D\u003cbr data-mce-fragment=\"1\"\u003eCylindrical refractive power (C)\u003cbr data-mce-fragment=\"1\"\u003eMeasurement range .............. 0 D to ±10.00 D\u003cbr data-mce-fragment=\"1\"\u003e (at VD = 12.0 mm)\u003cbr data-mce-fragment=\"1\"\u003eDisplay unit ........................... 0.01 D \/ 0.12 D \/ 0.25 D\u003cbr data-mce-fragment=\"1\"\u003eAstigmatism axial (A)\u003cbr data-mce-fragment=\"1\"\u003eMeasurement range .............. 0° to 180°\u003cbr data-mce-fragment=\"1\"\u003eDisplay unit ........................... 1°\u003cbr data-mce-fragment=\"1\"\u003eCorneal curvature measurement (K1, K2, AVG)\u003cbr data-mce-fragment=\"1\"\u003eMeasurement range .............. 5.00 mm to 11.00 mm \/ 30.68 D\u003cbr data-mce-fragment=\"1\"\u003e to 67.50 D (n=1.3375)\u003cbr data-mce-fragment=\"1\"\u003eDisplay unit ........................... 0.01 mm\u003cbr data-mce-fragment=\"1\"\u003eCorneal astigmatism and axis (C, A)\u003cbr data-mce-fragment=\"1\"\u003eMeasurement range (C) ........ 0 D to 10 D (n=1.3375)\u003cbr data-mce-fragment=\"1\"\u003eMeasurement range (A) ........ 0° to 180°\u003cbr data-mce-fragment=\"1\"\u003eMeasurement area cornea .... Ø 3.0 mm\u003cbr data-mce-fragment=\"1\"\u003e (at 8.00 mm corneal curvature)\u003cbr data-mce-fragment=\"1\"\u003ePD range ............................... 50 mm to 86 mm\u003cbr data-mce-fragment=\"1\"\u003eMinimum pupil diameter ....... Ø 2.0 mm\u003cbr data-mce-fragment=\"1\"\u003eVertex distance ..................... 0 mm - 16.0 mm\u003cbr data-mce-fragment=\"1\"\u003eMain unit\u003cbr data-mce-fragment=\"1\"\u003eBuilt-in printer ....................... Thermal printer\u003cbr data-mce-fragment=\"1\"\u003eOutput .................................. RS 232 C\u003cbr data-mce-fragment=\"1\"\u003eDisplay .................................. 5.7“ colour LCD display\u003cbr data-mce-fragment=\"1\"\u003eChin rest ................................ electr. controlled\u003cbr data-mce-fragment=\"1\"\u003eDimensions \u0026amp; electric Requirements\u003cbr data-mce-fragment=\"1\"\u003eDimensions WDH .................. 297 x 500 x 448 mm\u003cbr data-mce-fragment=\"1\"\u003eWeight ................................. Approx. 17.0 kg\u003cbr data-mce-fragment=\"1\"\u003eVoltage ................................. 100 VAC to 240 VAC\u003cbr data-mce-fragment=\"1\"\u003eFrequency ............................. 50\/60 Hz\u003cbr data-mce-fragment=\"1\"\u003ePower consumption .............. 80 VA to 100 VA\u003c\/span\u003e\u003c\/p\u003e","vendor":"optiusa","product_type":"","created_at":"2022-02-21T16:00:36-05:00","handle":"cx-800-auto-refractor-keratometer","updated_at":"2022-03-03T15:58:00-05:00","published_at":"2022-02-22T10:49:09-05:00","template_suffix":"rodenstock-instruments","status":"active","published_scope":"web","tags":"","admin_graphql_api_id":"gid:\/\/shopify\/Product\/7553620082903","variants":[{"id":42431512936663,"product_id":7553620082903,"title":"Default Title","price":"7296.00","sku":"","position":1,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Default Title","option2":null,"option3":null,"created_at":"2022-02-21T16:00:36-05:00","updated_at":"2022-02-22T13:39:45-05:00","taxable":true,"barcode":"","grams":19000,"image_id":null,"weight":19.0,"weight_unit":"kg","inventory_item_id":44525966885079,"inventory_quantity":0,"old_inventory_quantity":0,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42431512936663"}],"options":[{"id":9613121749207,"product_id":7553620082903,"name":"Title","position":1,"values":["Default Title"]}],"images":[{"id":36673445429463,"product_id":7553620082903,"position":1,"created_at":"2022-02-21T16:00:38-05:00","updated_at":"2022-02-21T16:00:38-05:00","alt":null,"width":1200,"height":600,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/product_570-914-1.jpg?v=1645477238","variant_ids":[],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36673445429463"}],"image":{"id":36673445429463,"product_id":7553620082903,"position":1,"created_at":"2022-02-21T16:00:38-05:00","updated_at":"2022-02-21T16:00:38-05:00","alt":null,"width":1200,"height":600,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/product_570-914-1.jpg?v=1645477238","variant_ids":[],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36673445429463"}},{"id":7550715560151,"title":"RSL 5000 Digital Slit Lamp Haag-Streit TypeTESTTESTTESTTEST","body_html":"\u003cspan id=\"ctl00_ContentPlaceHolder1_lblDescription\" data-mce-fragment=\"1\"\u003e• Elegant and classic design\u003cbr data-mce-fragment=\"1\"\u003e• Exceptional optics\u003cbr data-mce-fragment=\"1\"\u003e• 5-step magnification changer\u003cbr data-mce-fragment=\"1\"\u003e• Outstanding reliability\u003cbr data-mce-fragment=\"1\"\u003e• Extensive filter and accessory set\u003cbr data-mce-fragment=\"1\"\u003e• Integrated yellow filter\u003cbr data-mce-fragment=\"1\"\u003e• LED illumination\u003cbr data-mce-fragment=\"1\"\u003e• R-Type Tonometer available\u003c\/span\u003e","vendor":"optiusa","product_type":"","created_at":"2022-02-18T11:52:38-05:00","handle":"rsl-5000-digital-slit-lamp-haag-streit-type","updated_at":"2022-03-14T15:54:45-04:00","published_at":null,"template_suffix":"","status":"draft","published_scope":"web","tags":"","admin_graphql_api_id":"gid:\/\/shopify\/Product\/7550715560151","variants":[{"id":42430763466967,"product_id":7550715560151,"title":"Black","price":"999999.00","sku":"1","position":1,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Black","option2":null,"option3":null,"created_at":"2022-02-21T10:42:16-05:00","updated_at":"2022-02-21T10:42:16-05:00","taxable":true,"barcode":"","grams":0,"image_id":null,"weight":0.0,"weight_unit":"lb","inventory_item_id":44525216956631,"inventory_quantity":1,"old_inventory_quantity":1,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42430763466967"},{"id":42430763499735,"product_id":7550715560151,"title":"Red","price":"999999.00","sku":"2","position":2,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Red","option2":null,"option3":null,"created_at":"2022-02-21T10:42:16-05:00","updated_at":"2022-02-21T10:42:16-05:00","taxable":true,"barcode":"","grams":0,"image_id":null,"weight":0.0,"weight_unit":"lb","inventory_item_id":44525216989399,"inventory_quantity":0,"old_inventory_quantity":0,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42430763499735"},{"id":42430763532503,"product_id":7550715560151,"title":"Blue","price":"999999.00","sku":"3","position":3,"inventory_policy":"deny","compare_at_price":null,"fulfillment_service":"manual","inventory_management":"shopify","option1":"Blue","option2":null,"option3":null,"created_at":"2022-02-21T10:42:16-05:00","updated_at":"2022-02-21T10:42:16-05:00","taxable":true,"barcode":"","grams":0,"image_id":null,"weight":0.0,"weight_unit":"lb","inventory_item_id":44525217022167,"inventory_quantity":0,"old_inventory_quantity":0,"requires_shipping":true,"admin_graphql_api_id":"gid:\/\/shopify\/ProductVariant\/42430763532503"}],"options":[{"id":9609314336983,"product_id":7550715560151,"name":"Color","position":1,"values":["Black","Red","Blue"]}],"images":[{"id":36661145272535,"product_id":7550715560151,"position":1,"created_at":"2022-02-18T11:52:40-05:00","updated_at":"2022-02-18T11:52:40-05:00","alt":null,"width":1500,"height":750,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/1.jpg?v=1645203160","variant_ids":[],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36661145272535"},{"id":36661145239767,"product_id":7550715560151,"position":2,"created_at":"2022-02-18T11:52:40-05:00","updated_at":"2022-02-18T11:52:40-05:00","alt":null,"width":1500,"height":750,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/2.jpg?v=1645203160","variant_ids":[],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36661145239767"},{"id":36661145206999,"product_id":7550715560151,"position":3,"created_at":"2022-02-18T11:52:40-05:00","updated_at":"2022-02-18T11:52:40-05:00","alt":null,"width":1500,"height":750,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/3.jpg?v=1645203160","variant_ids":[],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36661145206999"}],"image":{"id":36661145272535,"product_id":7550715560151,"position":1,"created_at":"2022-02-18T11:52:40-05:00","updated_at":"2022-02-18T11:52:40-05:00","alt":null,"width":1500,"height":750,"src":"https:\/\/cdn.shopify.com\/s\/files\/1\/0627\/2152\/3927\/products\/1.jpg?v=1645203160","variant_ids":[],"admin_graphql_api_id":"gid:\/\/shopify\/ProductImage\/36661145272535"}}]}

**Comment by flgatormike** [comment]
When I get orders, I get a root element of "orders"
When I get products, I get a root element of "products"
When I get customers, I get a root element of "customers"
When I update a product, I get a root element of "product"
And those all work fine
But when I create a product, I get a root element of "product**s**" when it should be "product"
Any ideas?

**Comment by nozzlegear** [comment]
Honestly it's very strange behavior. Just to confirm, which version of ShopifySharp are you using? And have you extended the ProductService with your own class/override methods at all?

**Comment by flgatormike** [comment]
> Honestly it's very strange behavior. Just to confirm, which version of ShopifySharp are you using? And have you extended the ProductService with your own class/override methods at all?

No I haven't extended it. The store owner says it is "Store 2.0", could it be something with the way the store was configured?

**Comment by flgatormike** [comment]
So when I call the different ProductService methods, I took a look at the url that is being called in ExecuteRequestAsync. They are all fine except for CreateAsync which is posting to products.json.. the content looks fine, it is posting the product data but the uri doesn't look right. Not sure where that is being set?

ProductService > ListAsync
root element: products
uri: https://mydomain/admin/api/2021-10/products.json

ProductService > GetProduct
root element: product
uri: https://mydomain/admin/api/2021-10/products/7550715560151.json

ProductService > UpdateAsync
root element: product
uri: https://mydomain/admin/api/2021-10/products/7550715560151.json

ProductService > CreateAsync
root element: product
uri: **https://mydomain/admin/api/2021-10/products.json**

**Comment by flgatormike** [comment]
I guess that endpoint is correct for creating a new product, however mine is 2021-10 and the one in the shopify documentation is 2022-01, just a difference of 3 months, that couldn't be my problem could it?

![image](https://user-images.githubusercontent.com/47644612/159387091-0aa973ff-8118-4cfb-99e4-da67e247bb98.png)

**Comment by nozzlegear** [comment]
That shouldn't be an issue, the 2021-10 version we're using right now is still supported until October 2022. I just had a thought though! In the email you sent me earlier today, the shop website you're using looks like the "real" domain but it's not the shop's *.myshopify.com domain. 

I think Shopify is redirecting the request to the shop's *.myshopify.com domain by returning a 301 or 302 redirect. The HttpClient might be following that redirect but not preserving the POST method, instead turning it into a GET request which is why Shopify would return a list of products (because listing products and creating a new product both use the same `products.json` endpoint).

Can you try changing your code to use your shop's *.myshopify.com domain? 

var service = new ProductService("example.myshopify.com", accessToken);

--
Josh Harms

**Comment by flgatormike** [comment]
Hey, that worked! Thanks so much! It is strange that the update PUT worked with the other domain but the create POST didn't...

**Comment by nozzlegear** [comment]
A little note for myself here, I've added an FAQ to the readme that mentions these domain problems.

**Comment by pmartinciibo** [comment]
I am having a very similar issue, though I have made sure to use *.myshopify.com as the domain. When I try to create a product, I receive back an 'Object reference not set to an instance of an object.' that is being thrown at {T DeserializeWithNewtonsoft[T](System.String, System.String, System.Nullable`1[Newtonsoft.Json.DateParseHandling])}. GETing the products works correctly. This is the first time I have tried to create anything (POST for order, customer, product, etc.). Any ideas?
